### PR TITLE
[3.0] use ContainerAwareTrait

### PR DESCRIPTION
**Pull request by Tobion**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Finished version of #12595 

This adds the first deprecation in symfony 3. We cannot deprecate ContainerAware in 2.8 as the alternative, the trait, cannot be used with php 5.3. So there would not be an upgrade path for people using an older php version

### CODE REVIEW

**Comment by Tobion** [comment]
@symfony/deciders this is ready.

File: src/Symfony/Component/DependencyInjection/ContainerAware.php [hunk 1]
```diff
@@ -11,10 +11,14 @@
...
+@trigger_error('The '.__NAMESPACE__.'\ContainerAware class is deprecated since version 3.0 and will be removed in 4.0. Use ContainerAwareTrait with ContainerAwareInterface instead.', E_USER_DEPRECATED);
+
 /**
  * A simple implementation of ContainerAwareInterface.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ * 
+ * @deprecated since version 3.0, to be removed in 4.0. Use ContainerAwareTrait with ContainerAwareInterface instead.
  */
 abstract class ContainerAware implements ContainerAwareInterface
```
Comment (aitboudad) [inline src/Symfony/Component/DependencyInjection/ContainerAware.php]: we can use `ContainerAwareTrait` in this class too

Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ContainerAware.php]: good catch, changed

**Comment by aitboudad** [comment]
:+1:

**Comment by nicolas-grekas** [comment]
Shouldn't we deprecate in 2.8? IMHO yes, what do others think?

**Comment by Tobion** [comment]
@nicolas-grekas please read the description.

**Comment by WouterJ** [comment]
I think we should deprecate in 2.8:

 * Symfony 3 requires PHP 5.5
 * Symfony promises a very smooth (no changes required) upgrade process from 2.8 to 3.0 when having no deprecation notices in 2.8
 * This means applications have to run Symfony 2.8 on PHP 5.5 already
 * Which means they can use the trait (as it's available from PHP 5.4)

Actually, maybe we should even trigger a deprecation notice when PHP <5.5 in `Kernel#handle` or the like in Symfony 2.8.

**Comment by nicolas-grekas** [comment]
Of course there would be an upgrade path: first upgrade to php5.5, which is already required. I agree with @wouterj arguments (but for notices, they should not be triggered conditionally upon the php version)

**Comment by TomasVotruba** [comment]
:+1: Awesome

**Comment by Tobion** [comment]
IMO developers should be able to upgrade their 2.x projects to 2.8 while avoiding any deprecation message. People stuck on PHP 5.3 would not be able to do so.
Anyway, I'm ok with whatever approach we take. Time for @fabpot to decide :)

File: src/Symfony/Bridge/Doctrine/ManagerRegistry.php [hunk 2]
```diff
@@ -22,10 +22,7 @@
  */
 abstract class ManagerRegistry extends AbstractManagerRegistry implements ContainerAwareInterface
```
Comment (Taluu) [inline src/Symfony/Bridge/Doctrine/ManagerRegistry.php]: you should remove the implements here

**Comment by jakzal** [comment]
status: reviewed

:+1:

**Comment by fabpot** [comment]
IIRC, we agreed to not have any deprecation notices in 3.0. So, it's either 2.8 or 3.1.

**Comment by Tobion** [comment]
Lets just merge #16424, ill rebase this one then.

**Comment by nicolas-grekas** [comment]
While rebasing @Tobion you could also remove the uses of the VarDumperTestCase also (replaced by VarDumperTestTrait). Or I can do it, as you want.

**Comment by nicolas-grekas** [comment]
Sibling PR on 2.8 to not break 3.0 compat: #16466

**Comment by Tobion** [comment]
Rebased and I found two more classes (commands in the TwigBundle) where we can use the ContainerAwareTrait. I made those commands final because they is not meant to be extended. By using the trait, which uses `protected` visibility for the container property, it would otherwise make the container property part of of BC promise.
I also changed the VarDumper to use its trait.

File: src/Symfony/Component/VarDumper/Tests/Test/VarDumpTestTraitTest.php [hunk 3]
```diff
@@ -9,8 +9,32 @@
...
 
-// Skipping trait tests for PHP < 5.4
-if (version_compare(PHP_VERSION, '5.4.0-dev', '>=')) {
-    require 'VarDumpTestTraitRequire54.php';
-}
+namespace Symfony\Component\VarDumper\Tests\Test;
+
+use Symfony\Component\VarDumper\Test\VarDumperTestTrait;
+
+class VarDumperTestTraitTest extends \PHPUnit_Framework_TestCase
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Tests/Test/VarDumpTestTraitTest.php]: a mistake that was here before but could be fixed here: the name of the file is wrong (VarDump**er**)

Comment (Tobion) [inline src/Symfony/Component/VarDumper/Tests/Test/VarDumpTestTraitTest.php]: done

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by fabpot** [comment]
Tests are broken.

**Comment by nicolas-grekas** [comment]
The hard coded line info (63 to 63) needs to be updated (65 to 65)

**Comment by Tobion** [comment]
Fixed tests

**Comment by nicolas-grekas** [comment]
Thank you @Tobion.

**Pull request merged**